src/core/binding.c

```diff
@@ -560,46 +560,100 @@ _IRQL_requires_max_(DISPATCH_LEVEL)
 BOOLEAN
 QuicBindingAddSourceConnectionID(
     _In_ QUIC_BINDING* Binding,
-    _In_ QUIC_CID_HASH_ENTRY* SourceCid
+    _In_ QUIC_CID_SLIST_ENTRY* SourceCid
     )
 {
     return QuicLookupAddLocalCid(&Binding->Lookup, SourceCid, NULL);
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+BOOLEAN
+QuicBindingAddAllSourceConnectionIDs(
+    _In_ QUIC_BINDING* Binding,
+    _In_ QUIC_CONNECTION* Connection
+    )
+{
+    for (CXPLAT_SLIST_ENTRY* Link = Connection->SourceCids.Next;
+        Link != NULL;
+        Link = Link->Next) {
+
+        QUIC_CID_SLIST_ENTRY* Entry =
+            CXPLAT_CONTAINING_RECORD(
+                Link,
+                QUIC_CID_SLIST_ENTRY,
+                Link);
+        if (!QuicBindingAddSourceConnectionID(Binding, Entry)) {
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
 _IRQL_requires_max_(DISPATCH_LEVEL)
 void
 QuicBindingRemoveSourceConnectionID(
     _In_ QUIC_BINDING* Binding,
-    _In_ QUIC_CID_HASH_ENTRY* SourceCid,
-    _In_ CXPLAT_SLIST_ENTRY** Entry
+    _In_ QUIC_CID_HASH_ENTRY* SourceCid
     )
 {
-    QuicLookupRemoveLocalCid(&Binding->Lookup, SourceCid, Entry);
+    QuicLookupRemoveLocalCid(&Binding->Lookup, SourceCid);
 }
 
 _IRQL_requires_max_(DISPATCH_LEVEL)
 void
-QuicBindingRemoveConnection(
+QuicBindingRemoveAllSourceConnectionIDs(
     _In_ QUIC_BINDING* Binding,
     _In_ QUIC_CONNECTION* Connection
     )
 {
-    if (Connection->RemoteHashEntry != NULL) {
-        QuicLookupRemoveRemoteHash(&Binding->Lookup, Connection->RemoteHashEntry);
+    CXPLAT_SLIST_ENTRY EntriesToFree = {0};
+
+    for (CXPLAT_SLIST_ENTRY* Link = Connection->SourceCids.Next;
+        Link != NULL;
+        Link = Link->Next) {
+
+        QUIC_CID_SLIST_ENTRY* Entry =
+            CXPLAT_CONTAINING_RECORD(
+                Link,
+                QUIC_CID_SLIST_ENTRY,
+                Link);
+
+        CXPLAT_SLIST_ENTRY** Link1 = &Entry->HashEntries.Next;
+        while (*Link1 != NULL) {
+            QUIC_CID_HASH_ENTRY* Entry1 = 
+                CXPLAT_CONTAINING_RECORD(
+                    *Link1,
+                    QUIC_CID_HASH_ENTRY,
+                    Link);
+            if (Entry1->Binding == Binding) {
+                QuicBindingRemoveSourceConnectionID(Binding, Entry1);
+                *Link1 = (*Link1)->Next;
+                CxPlatListPushEntry(&EntriesToFree, &Entry1->Link);
```

Comment:
Why have a separate list that you have to iterate again later to free? Just free here directly.